### ISSUE TITLE
Integrating ExoPlayer with system equalizer

### ISSUE THREAD

**Issue by mkaflowski** [google/ExoPlayer#3058]
What is the best way to integrate ExoPlayer with system equalizer? I do it in onAudioSessionId but I get effects delayed. 

```           
@Override
public void onAudioSessionId(int audioSessionId) {
    final Intent intent = new Intent(AudioEffect.ACTION_OPEN_AUDIO_EFFECT_CONTROL_SESSION);
    intent.putExtra(AudioEffect.EXTRA_AUDIO_SESSION, player.getAudioSessionId());
    intent.putExtra(AudioEffect.EXTRA_PACKAGE_NAME, context.getPackageName());
    context.sendBroadcast(intent);
}
```

Is there any way to set my own onAudioSessionId (it could be workaround for me)?

**Comment by ojw28** [comment]
@andrewlewis - Could/should we should generate the first audio session ID up front (possibly as soon as the player is instantiated), so it can be queried by the app before playback is started?

**Comment by mkaflowski** [comment]
Does setting my own session Id generated before could do the job?

**Comment by andrewlewis** [comment]
@mkaflowski Does the delay go away if you call `setPlayWhenReady(true)` only once you have the audio session and the equalizer is active? The player should get an audio session while buffering if audio is enabled.

We can generate a session ID up front but only from API 21 using `AudioManager.generateAudioSessionId()` (maybe it should happen lazily in a getter on `SimpleExoPlayer`). Before API 21 I think it would be necessarily to instantiate and initialize an `AudioTrack`, which is not a good option because it uses shared memory and audio might be missing/disabled.

**Comment by mkaflowski** [comment]
I think thant delay is the system issue. It takes a while from sending AudioEffect.ACTION_OPEN_AUDIO_EFFECT_CONTROL_SESSION to activating it for my session (I test on Sony phones). Sometimes it is couple senconds, most likely about 0.5s but it is hearable when you choose some track from the list and start playing right after initializing player.
I have done some research on github project and some developers just sets ession id whitch has the effected invoked earlier from other MediaPlayer instances by themeself but I haven't found that possibility in Exoplayer.
I think about workaround and I will give feedback if I come up with something. Some apps like Spotify use exoplayer and don't have this problem so I think that there is some solution.

**Comment by andrewlewis** [comment]
Did you find a workaround for this? As suggested above, I wonder whether you can just prepare the player in the paused state, at which point you should get an audio session ID, then unpause the player only once the equalizer is set up.

**Comment by mkaflowski** [comment]
Not yet. I have beed waiting for a release with DynamicConcatenatingMediaSource and will try do something with that.

**Comment by andrewlewis** [comment]
As this thread describes a possible workaround, I'm going to close the issue. If an enhancement is still needed in this area please file a new issue and describe the behavior you're looking for. Thanks.

**Comment by jakoss** [comment]
I have similar issue. I have to handle situation when player is not playing (app is just opened, ExoPlayer is created but user didn't start any playback. In this app state i need to open system Equalizer, but there is no AudioSessionId set yet. Is there an option to generate manually AudioSessionId and then fetch it to Player once it's playing?

**Comment by andrewlewis** [comment]
@Nekromancer From API 21 onward you can generate a session ID using `C.generateAudioSessionIdV21`, and pass the generated session ID to a `DefaultAudioSink` you create yourself and pass to the audio renderer. Does that work for your use case? If so we should probably add a setter on `SimpleExoPlayer` to make this easier to do.

**Comment by jakoss** [comment]
Well, i'm not really sure how to do this. I'm using `DefaultRenderersFactory` that i pass to `ExoPlayerFactory.newSimpleInstance` along with context and track selector. As far as i can tell - `DefaultAudioSink` gets created in `createRenderers` method, but i'm not sure how to pass my `AudioSink` into this.

Setter on `SimpleExoPlayer` would be the best option for us

**Comment by master255** [comment]
I found a way to set Audiosessionid to DefaultAudioSink in DefaultRenderersFactory, but then a bug occurs. There is a bug anyway. After playing mkv in 4k using ffmpeg audio decoder, an event arrives that wipes the audio session id. The correct audio session remains in MediaCodecAudioRenderer, but SimpleExoPlayer returns 0. For [example.](http://master255.org/res/%d0%a4%d0%b8%d0%bb%d1%8c%d0%bc%d1%8b%20(TOP)/4k%20%d1%84%d0%b8%d0%bb%d1%8c%d0%bc%d1%8b/%d0%92%d1%81%d0%bf%d0%be%d0%bc%d0%bd%d0%b8%d1%82%d1%8c%20%d0%b2%d1%81%d1%91%20(2012,%20WEB-DL%202160).mkv)
The setForegroundMode value is enabled.
How to get the right audio session after watching a 4k movie?

**Comment by master255** [comment]
And it’s very wrong to change the audio session. I would like to use one for the entire time using the player. This will reduce the load on the processor and increase the battery life. Is this fundamentally possible?

**Comment by master255** [comment]
A temporary bug fix may be:
SimpleExoplayer.java 1345
add check:  || sessionId < 1
But this does not fix the generation of new audio sessions and the high processor load due to this. This is a very bad practice.

**Comment by andrewlewis** [comment]
Sorry for the delay. Yes, I think we are currently clearing the `audioSessionId` when renderers are disabled, even when it was set by the app.

I think we can use this bug for tracking these related tasks:
- Providing a convenience method to generate an audio session ID. TBD how we deal with the fact that it's only possible to generate an audio session ID from API 21 onwards.
- Retaining the audio session ID in `DefaultAudioSink` across the player lifetime.
- Improving documentation about using platform audio effects (also tracked by [Internal: b/145397915].

**Comment by vyagrah** [comment]
where is this `onAudioSessionId`?
is it deprecated?
I used `onAudioSessionIdChanged()` meathods of `AnalyticsListener `and `Player.Listener` but they never get called.

**Comment by ojw28** [comment]
We made some improvements to the way audio session IDs are generated in 2.13.0. The corresponding entries in the release notes are:

*   Fix handling of audio session IDs ([#8190](https://github.com/google/ExoPlayer/issues/8190)). `SimpleExoPlayer` now generates an audio session ID on construction, which can be immediately queried by calling `SimpleExoPlayer.getAudioSessionId`. The audio session ID will only change if application code calls `SimpleExoPlayer.setAudioSessionId`.
* Replace `onAudioSessionId` with `onAudioSessionIdChanged` in `AudioListener` and `AnalyticsListener`. Note that `onAudioSessionIdChanged` is called in fewer cases than `onAudioSessionId` was called, due to the improved handling of audio session IDs as described above.

So, you found the right method for being notified when the audio session ID is changed, but since it's not changing the method is not being called. You can use `getAudioSessionId` immediately after player construction to query the audio session ID that's being used.
